contracts/upgradeable_contracts/arbitrary_message/MessageDelivery.sol

```diff
@@ -2,22 +2,48 @@ pragma solidity 0.4.24;
 
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 import "./BasicAMB.sol";
+import "./VersionableAMB.sol";
+import "../../libraries/Bytes.sol";
 
-contract MessageDelivery is BasicAMB {
+contract MessageDelivery is BasicAMB, VersionableAMB {
     using SafeMath for uint256;
 
-    function requireToPassMessage(address _contract, bytes _data, uint256 _gas) public {
+    /**
+    * @dev Requests message relay to the opposite network
+    * @param _contract executor address on the other side
+    * @param _data calldata passed to the executor on the other side
+    * @param _gas gas limit used on the other network for executing a message
+    */
+    function requireToPassMessage(address _contract, bytes _data, uint256 _gas) public returns (bytes32) {
         require(_gas >= getMinimumGasUsage(_data) && _gas <= maxGasPerTx());
-        emitEventOnMessageRequest(abi.encodePacked(msg.sender, _contract, _gas, uint8(0x00), _data));
+
+        uint256 chainId = _chainId();
+        bytes20 bridgeId = bytes20(keccak256(abi.encodePacked(chainId, address(this))));
+        uint64 nonce = _nonce();
+        _setNonce(nonce + 1);
+
+        bytes32 messageId = Bytes.bytesToBytes32(abi.encodePacked(ENCODED_BRIDGE_VERSION, bridgeId, nonce));
+        bytes memory eventData = abi.encodePacked(
+            messageId,
+            chainId,
+            msg.sender,
+            _contract,
+            uint32(_gas),
+            uint8(0x00),
+            _data
+        );
+
+        emitEventOnMessageRequest(messageId, eventData);
```

Comment:
It looks like the `messageId` parameter is present in the two parameters of the event, in the first dedicated field and as part of the encodedData. Is this the expected behavior? I guess the tradeoff of this is the extra gas expended on the event vs the extra work of the validators to include the `messageId` into the message when processing the events

Reply:
In all events where `messageId` is present, it is an indexed field. So it can be used for searching of all events related for one particular message.

I don't know if it is useful at this moment, but I think it can be helpful in the future.